ldms/python/ldms.pyx

```diff
@@ -3056,6 +3056,45 @@ cdef class Xprt(object):
             rc = self._conn_rc
             raise ConnectionError(rc, "Connect error: {}".format(ERRNO_SYM(rc)))
 
+    def tconnect(self, host, port=411, timeout=5):
+        """X.tconnect(host, port=411, timeout=5)
+
+        Connect to the remote LDMS peer
+
+        Arguments:
+        - host (str): The hostname (or IP address in string) to connect to.
+        - port (int): The peer port number.
+        - timeout (int): The amount of time to wait before timing out connection
+
+        'tconnect` is a blocking function, i.e. it will not
+        return until the either connection resolves in success or failure, or the
+        connection times out. In the case of success, it simply returns. Otherwise,
+        a ConnectionError is raised.
+        """
+        cdef int rc
+        cdef timespec tout_spec
+        tout_spec = {}
+        clock_gettime(CLOCK_REALTIME, &tout_spec)
+        tout_spec.tv_sec = tout_spec.tv_sec + timeout
+        self._conn_cb = None
+        self._conn_cb_arg = None
+
+        rc = ldms_xprt_connect_by_name(self.xprt, BYTES(host), BYTES(port),
+                                       xprt_cb, <void*>self)
+        if rc != 0:
+            # synchronously failed, self.xprt is no good. Need to "put" it down.
+            ldms_xprt_put(self.xprt)
+            self.xprt = NULL
+            raise ConnectionError(rc, "ldms_xprt_timed_connect_by_name() error: {}" \
+                                      .format(ERRNO_SYM(rc)))
+        with nogil:
+            sem_timedwait(&self._conn_sem, &tout_spec)
```

Comment:
I think you also need to add a logic in `xprt_cb()` function to detect the events on the connecting endpoint that has already timed out and handle them properly (e.g. not posting a semaphore as the matching `sem_timedwait` has already timed out).